src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -70915,5 +70937,191 @@ static void G<T>(T? x, object y)
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c2").WithLocation(9, 9),
+                // (15,9): warning CS8602: Possible dereference of a null reference.
+                //         c2.ToString(); // Warn 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c2").WithLocation(15, 9),
+                // (19,31): warning CS8604: Possible null reference argument for parameter 'c' in 'C C.GetC(C c)'.
+                //         c1 ??= (c2 = c3).GetC(c1); // Warn 3
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "c1").WithArguments("c", "C C.GetC(C c)").WithLocation(19, 31));
+        }
+
+        [WorkItem(30140, "https://github.com/dotnet/roslyn/issues/30140")]
+        [Fact]
+        public void NullCoalescingAssignment_Contravariant()
+        {
+            var source = @"
+class C
+{
+#nullable disable
+    C GetC() => null;
+#nullable enable
+    void M(C? c1, C c2)
+    {
+        // nullable + oblivious = oblivious
+#nullable disable
+        C c3
+#nullable enable
+                  = (c1 ??= GetC());
+        _ = c1/*T:C*/;
+        _ = c3/*T:C*/;
+
+        // oblivious + nullable = nullable
+        // This one is a little weirder. The ??= on c3 forces it into either nullable or not nullable, so we must
+        // pick the nullable version, meaning the resulting expression is actually nullable + nullable = nullable.
+        c1 = null;
+        var c4 = (c3 ??= c1);
+        _ = c3/*T:C?*/;
+        _ = c4/*T:C?*/;
+
+        // oblivious + not nullable = not nullable
+        c3 = GetC();
+        var c5 = (c3 ??= c2);
+        _ = c3/*T:C!*/;
+        _ = c5/*T:C!*/;
+
+        // not nullable + oblivious = not nullable
+        var c6 = (c2 ??= GetC());
+        _ = c2/*T:C!*/;
+        _ = c6/*T:C!*/;
+    }
+}";
+
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (30,19): hidden CS8607: Expression is probably never null.
+                //         var c6 = (c2 ??= GetC());
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "c2").WithLocation(32, 19));
+            comp.VerifyTypes();
+        }
+
+
+        [WorkItem(30140, "https://github.com/dotnet/roslyn/issues/30140")]
+        [Fact]
+        public void NullCoalescingAssignment_LeftNotTracked()
+        {
+            var source = @"
+class C
+{
+#nullable disable
+    C GetC() => null;
+#nullable enable
+    void M(C?[] c1, C c2)
+    {
+        c1[0] ??= c2;
+        c1[0].ToString(); // Warn
+    }
+}";
+
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (10,9): warning CS8602: Possible dereference of a null reference.
+                //         c1[0].ToString(); // Warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1[0]").WithLocation(10, 9));
+        }
```

Comment:
>} [](start = 8, length = 1)

It looks like none of the added tests target unconstrained type parameters. Consider "cloning" NullabilityOfTypeParameters_* tests that target coalesce expression.